rust/simd/README.md

```diff
@@ -0,0 +1,215 @@
+[...]
+Deployed canisters.
+URLs:
+   Backend canister via Candid interface:
+      mat_mat_mul: http://127.0.0.1/?canisterId=...
+```
+
+- #### Step 2.5: Compare the amount of instructions used for different matrix multiplication implementations
+
+```sh
+dfx canister call mat_mat_mul generic_f32
+dfx canister call mat_mat_mul auto_vectorization_f32
+```
+
+Example output:
+
+```sh
+% dfx canister call mat_mat_mul generic_f32
+(31_484_435 : nat64)
+% dfx canister call mat_mat_mul auto_vectorization_f32
+(15_249_432 : nat64)
+```
+
+Rust auto-vectorization again demonstrates its power in this example!
+The auto-vectorized version of the integer matrix multiplication achieves
+more than a 2x speedup compared to the original code.
+
+However, the slower performance of the generic floating-point
+matrix multiplication is due to a known issue with NaN canonicalization in `wasmtime`.
+This issue [has been fixed](https://github.com/bytecodealliance/wasmtime/commit/72a3b8b99d7c0343bacb7cd2cff3151b0144179d)
+by DFINITY, but not yet released at the time of writing.
+
+## Further learning
+
+1. Have a look at the locally running dashboard. The URL is at the end of the `dfx start` command: `Dashboard: http://localhost/...`
+2. Check out `mat_mat_mul` canister Candid user interface. The URLs are at the end of the `dfx deploy` command: `mat_mat_mul: http://127.0.0.1/?canisterId=...`
+
+### Canister interface
+
+The `mat_mat_mul` canister provide the following interface:
+
+- `generic_f32` &mdash; returns the number of instructions used for a loop performing element-wise multiplication of `224x4` tiles of matrices `A` and `B` using generic Rust operators.
+- `auto_vectorization_f32` &mdash; returns the number of instructions used for a loop performing element-wise multiplication of `224x4` tiles of matrices `A` and `B` using loop auto-vectorization for generic Rust operators.
+- `simd_f32` &mdash; returns the number of instructions used for a loop performing element-wise multiplication of `224x4` tiles of matrices `A` and `B` using WebAssembly SIMD instructions.
+- `generic_u32` &mdash; returns the number of instructions used for a loop performing element-wise multiplication of `224x4` tiles of integer matrices `A` and `B` using generic Rust operators.
+- `auto_vectorization_u32` &mdash; returns the number of instructions used for a loop performing element-wise multiplication of `224x4` tiles of integer matrices `A` and `B` using loop auto-vectorization for generic Rust operators.
+
+Example usage:
+
+```sh
+dfx canister call mat_mat_mul generic_f32
+```
+
+## Conclusion
+
+WebAssembly SIMD instructions unlock new possibilities for the Internet Computer,
+particularly in Machine Learning and Artificial Intelligence dApps. This example
+demonstrates potential 10x speedups for matrix multiplication with minimal effort
+using just Rust's loop auto-vectorization.
+
+As shown in Example 2, integer operations also benefit, although with a more modest
+"2x" speedup.
```

Comment:
I think we can get higher speedup for integer using 8-bit integers and doing 16x16 tiles. Just fyi, nothing actionable here.